lib/libspl/asm-generic/atomic.c

```diff
@@ -313,19 +251,15 @@ ATOMIC_AND_NV(64, uint64_t)
 
 
 /*
- *  If *arg1 == arg2, set *arg1 = arg3; return old value
+ *  If *tgt == exp, set *tgt = des; return old value
  */
 
 #define	ATOMIC_CAS(name, type) \
-	type atomic_cas_##name(volatile type *target, type arg1, type arg2) \
+	type atomic_cas_##name(volatile type *target, type exp, type des) \
 	{								\
-		type old;						\
-		VERIFY3S(pthread_mutex_lock(&atomic_lock), ==, 0);	\
-		old = *target;						\
-		if (old == arg1)					\
-			*target = arg2;					\
-		VERIFY3S(pthread_mutex_unlock(&atomic_lock), ==, 0);	\
-		return (old);						\
+		__atomic_compare_exchange_n(target, &exp, des, B_FALSE,	\
+		    __ATOMIC_SEQ_CST, __ATOMIC_SEQ_CST);		\
+		return (exp);						\
```

Comment:
This doesn't look quite right.  The expected behavior for `atomic_cas_*` is to always return the original target value regardless of if the value was updated, not to return the passed expected value.

Reply:
The [description](https://gcc.gnu.org/onlinedocs/gcc/_005f_005fatomic-Builtins.html) for this is absolute *hell*, but if you look really closely around the middle bit, you'll see the following:
> If they are not equal, the operation is a read and the current contents of \*ptr are written into \*expected.

Reply:
Well that's clear as mud, can you add a small comment for future readers of these functions.

Reply:
Fair enough, added explanation to the comment above the macro.